vowpalwabbit/cb_to_cb_adf.cc

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "cb_to_cb_adf.h"
+
+#include "vw_versions.h"
+#include "reductions.h"
+#include "learner.h"
+#include "vw.h"
+#include "cbify.h"
+#include "cb_algs.h"
+#include "cb_explore.h"
+#include "cb_label_parser.h"
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+
+struct cb_to_cb_adf
+{
+  parameters* weights;
+  cbify_adf_data adf_data;
+  bool explore_mode;
+};
+
+template <bool is_learn>
+void predict_or_learn(cb_to_cb_adf& data, multi_learner& base, example& ec)
+{
+  data.adf_data.copy_example_to_adf(*data.weights, ec);
+
+  if (is_learn && !CB::is_test_label(&ec.l.cb))
+  {
+    uint32_t chosen_action = ec.l.cb.costs[0].action - 1;
+    if (chosen_action < data.adf_data.num_actions)
+    {
+      // TODO: do i need a guard here?
+      CB::label ld = data.adf_data.ecs[chosen_action]->l.cb;
+      data.adf_data.ecs[chosen_action]->l.cb = ec.l.cb;
+      base.learn(data.adf_data.ecs);
+      data.adf_data.ecs[chosen_action]->l.cb = ld;
+
+      CB::default_label(&data.adf_data.ecs[chosen_action]->l.cb);
```

Comment:
we are setting `data.adf_data.ecs[chosen_action]->l.cb` twice? one on line 40 and then on line 42 to feault label?

Reply:
I'll take a look on these 4 lines again.

Reply:
good catch, the second one should not be needed